pinot-broker/src/main/java/org/apache/pinot/broker/failuredetector/FailureDetectorFactory.java

```diff
@@ -0,0 +1,56 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.failuredetector;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.pinot.common.metrics.BrokerMetrics;
+import org.apache.pinot.spi.env.PinotConfiguration;
+import org.apache.pinot.spi.utils.CommonConstants.Broker;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class FailureDetectorFactory {
+  private FailureDetectorFactory() {
+  }
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(FailureDetectorFactory.class);
+
+  public static FailureDetector getFailureDetector(PinotConfiguration config, BrokerMetrics brokerMetrics) {
+    String className = config.getProperty(Broker.FailureDetector.CONFIG_OF_CLASS_NAME);
+    if (StringUtils.isEmpty(className)) {
+      LOGGER.info("Class name is not configured, falling back to NoOpFailureDetector");
+      return new NoOpFailureDetector();
+    } else {
+      LOGGER.info("Initializing failure detector with class: {}", className);
+      try {
+        FailureDetector failureDetector =
+            (FailureDetector) Class.forName(className).getDeclaredConstructor().newInstance();
```

Comment:
This isn’t a good configuration interface because it coupled users with the implementation. Instead have a configuration name and map it to an implementation class.

Reply:
(You can’t even move the class to another package without breaking users as things stand…)

Reply:
I don't fully follow the comment here, but changed it to use `PluginManager.get().createInstance(className)` so that we can take more plug-in classes. Does the problem still exist within the `PluginManager`?

Reply:
The point is exposing the class name at all prevents from changing the class name, which isn’t good. If you have a separate configuration setting and map that to the class you want to use, without exposing the class name, you can refactor if you want.

Reply:
I feel changing the model of how user plug in a custom implementation is out of the scope of this PR (currently we plug in custom implementation via `PluginManager` and class name). We may revisit this in a separate discussion just for the plug in model

Reply:
I agree that the plugin manager is worthy of a much bigger discussion, but that's not what's being discussed here.

You're adding a new configuration option associated with the key `Broker.FailureDetector.CONFIG_OF_CLASS_NAME`. That can either be a class name, **which means you can't change that class name ever again**, or it can be a value like "NONE", "DEFAULT" which you then _map_ to a class name to provide to the plugin manger, which allows non-breaking refactoring. By Exposing the class name, your implementation has become your interface.

Reply:
Added a config for the failure detector type so that the built-in detectors can be configured using the type